packages/browser/src/profiling/lifecycleMode/traceLifecycleProfiler.ts

```diff
@@ -0,0 +1,312 @@
+import type { JSSelfProfiler } from '../jsSelfProfiling';
+import { createProfileChunkPayload, startJSSelfProfile, validateProfileChunk } from '../utils';
+
+const CHUNK_INTERVAL_MS = 60_000;
+
+/**
+ * Browser trace-lifecycle profiler (UI Profiling / Profiling V2):
+ * - Starts when the first sampled root span starts
+ * - Stops when the last sampled root span ends
+ * - While running, periodically stops and restarts the JS self-profiling API to collect chunks
+ *
+ * Profiles are emitted as standalone `profile_chunk` envelopes either when:
+ * - there are no more sampled root spans, or
+ * - the 60s chunk timer elapses while profiling is running.
+ */
+export class BrowserTraceLifecycleProfiler {
+  private _client: Client | undefined;
+  private _profiler: JSSelfProfiler | undefined;
+  private _chunkTimer: ReturnType<typeof setTimeout> | undefined;
+  private _activeRootSpanCount: number;
+  // For keeping track of active root spans
+  private _activeRootSpanIds: Set<string>;
+  private _profilerId: string | undefined;
+  private _isRunning: boolean;
+  private _sessionSampled: boolean;
+
+  public constructor() {
+    this._client = undefined;
+    this._profiler = undefined;
+    this._chunkTimer = undefined;
+    this._activeRootSpanCount = 0;
+    this._activeRootSpanIds = new Set<string>();
+    this._profilerId = undefined;
+    this._isRunning = false;
+    this._sessionSampled = false;
+  }
+
+  /**
+   * Initialize the profiler with client and session sampling decision computed by the integration.
+   */
+  public initialize(client: Client, sessionSampled: boolean): void {
+    DEBUG_BUILD && debug.log("[Profiling] Initializing profiler (lifecycle='trace').");
+
+    this._client = client;
+    this._sessionSampled = sessionSampled;
+
+    client.on('spanStart', span => {
+      if (span !== getRootSpan(span)) {
+        return;
+      }
+      if (!this._sessionSampled) {
+        DEBUG_BUILD && debug.log('[Profiling] Session not sampled because of negative sampling decision.');
+        return;
+      }
+      // Only count sampled root spans
+      if (!span.isRecording()) {
+        DEBUG_BUILD && debug.log('[Profiling] Discarding profile because root span was not sampled.');
+        return;
+      }
+
+      const rootSpanJSON = spanToJSON(span);
+      const spanId = rootSpanJSON.span_id as string | undefined;
+      if (!spanId) {
+        return;
+      }
+      if (this._activeRootSpanIds.has(spanId)) {
+        return;
+      }
+      this._activeRootSpanIds.add(spanId);
+
+      const wasZero = this._activeRootSpanCount === 0;
+      this._activeRootSpanCount++; // Increment before eventually starting the profiler
+      DEBUG_BUILD &&
+        debug.log(
+          `[Profiling] Root span ${rootSpanJSON.description} started. Active root spans:`,
+          this._activeRootSpanCount,
+        );
+      if (wasZero) {
+        this.start();
+      }
+    });
+
+    client.on('spanEnd', span => {
```

Comment:
We are relying on the SDK to correctly call `spanEnd`, which means that if there is a bug or crash any of our or the SDK's span logic, there is risk of indefinitely running the profiler (iirc this actually happened in our python SDK). This can be problematic, as this executed on the client, meaning there is not kill switch to actually turn it off. 

Would you mind checking if there is some timeout related logic in the node SDK here, I believe we may have had something to prevent this